File: apps/desktop/src/api/recipes.ts
import { invoke } from "@tauri-apps/api/core";
import type { RecipeCostingGraph } from "@pastry-management/core";
import type { CostSnapshotSummary, RecipeDetail, RecipeInput, RecipeSummary } from "./types";

/**
 * Typed wrappers around the `commands::recipes` Tauri commands. This is the only module that
 * should call `invoke()` for recipes — components call these functions instead.
 *
 * Note: Tauri camelCases top-level command argument names for the JS side (e.g. Rust's
 * `include_archived` becomes `includeArchived` here), but does NOT rename fields inside a struct
 * argument like `RecipeInput` — those keep their exact snake_case Rust field names since that
 * struct has no `#[serde(rename_all = ...)]`.
 */

export function listRecipes(includeArchived: boolean): Promise<RecipeSummary[]> {
  return invoke("list_recipes", { includeArchived });
}

export function getRecipe(id: number): Promise<RecipeDetail> {
  return invoke("get_recipe", { id });
}

export function createRecipe(input: RecipeInput): Promise<RecipeDetail> {
  return invoke("create_recipe", { input });
}

export function updateRecipe(id: number, input: RecipeInput): Promise<RecipeDetail> {
  return invoke("update_recipe", { id, input });
}

export function archiveRecipe(id: number): Promise<void> {
  return invoke("archive_recipe", { id });
}

export function reactivateRecipe(id: number): Promise<void> {
  return invoke("reactivate_recipe", { id });
}

export function deleteRecipe(id: number): Promise<void> {
  return invoke("delete_recipe", { id });
}

/** Pass `newName: null` to get the backend's default "{name} (Copy)" behavior. */
export function duplicateRecipe(id: number, newName: string | null): Promise<RecipeDetail> {
  return invoke("duplicate_recipe", { id, newName });
}

export function getRecipeCostingGraph(id: number): Promise<RecipeCostingGraph> {
  return invoke("get_recipe_costing_graph", { id });
}

/**
 * Records an audit-trail snapshot of a cost calculation the user explicitly chose to save. Pass
 * `JSON.stringify(breakdown.pricingStrategyUsed)` and `JSON.stringify(breakdown)` for the last two
 * string arguments; the two money arguments must be whole integers (the costing engine's output is
 * already rounded to whole micros).
 */
export function saveRecipeCostSnapshot(
  recipeId: number,
  pricingStrategySummary: string,
  totalCostMicros: number,
  costPerYieldUnitMicros: number,
  breakdownJson: string,
): Promise<number> {
  return invoke("save_recipe_cost_snapshot", {
    recipeId,
    pricingStrategySummary,
    totalCostMicros,
    costPerYieldUnitMicros,
    breakdownJson,
  });
}

export function listRecipeCostSnapshots(recipeId: number): Promise<CostSnapshotSummary[]> {
  return invoke("list_recipe_cost_snapshots", { recipeId });
}

File: apps/desktop/src-tauri/src/commands/recipes.rs
use tauri::State;

use crate::auth::SessionState;
use crate::db::repositories::recipes::{
    self, CostSnapshotSummary, CostingRawMaterial, RecipeCostingGraph, RecipeDetail, RecipeInput,
    RecipeSummary,
};
use crate::error::{AppError, AppResult};
use crate::DbState;

fn current_user_id(session: &State<SessionState>) -> AppResult<Option<i64>> {
    Ok(session
        .0
        .lock()
        .map_err(|e| AppError::new(e.to_string()))?
        .as_ref()
        .map(|s| s.user.id))
}

#[tauri::command]
pub fn list_recipes(db: State<DbState>, include_archived: bool) -> AppResult<Vec<RecipeSummary>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::list(&conn, include_archived)
}

#[tauri::command]
pub fn get_recipe(db: State<DbState>, id: i64) -> AppResult<RecipeDetail> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::get(&conn, id)
}

#[tauri::command]
pub fn create_recipe(
    db: State<DbState>,
    session: State<SessionState>,
    input: RecipeInput,
) -> AppResult<RecipeDetail> {
    let user_id = current_user_id(&session)?;
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::create(&conn, input, user_id)
}

#[tauri::command]
pub fn update_recipe(
    db: State<DbState>,
    session: State<SessionState>,
    id: i64,
    input: RecipeInput,
) -> AppResult<RecipeDetail> {
    let user_id = current_user_id(&session)?;
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::update(&conn, id, input, user_id)
}

#[tauri::command]
pub fn archive_recipe(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::set_active(&conn, id, false)
}

#[tauri::command]
pub fn reactivate_recipe(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::set_active(&conn, id, true)
}

#[tauri::command]
pub fn delete_recipe(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::delete(&conn, id)
}

#[tauri::command]
pub fn duplicate_recipe(
    db: State<DbState>,
    id: i64,
    new_name: Option<String>,
) -> AppResult<RecipeDetail> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::duplicate(&conn, id, new_name)
}

#[tauri::command]
pub fn get_recipe_costing_graph(db: State<DbState>, id: i64) -> AppResult<RecipeCostingGraph> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::get_costing_graph(&conn, id)
}

#[tauri::command]
pub fn list_raw_material_costing(db: State<DbState>) -> AppResult<Vec<CostingRawMaterial>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::get_all_raw_materials_costing(&conn)
}

#[tauri::command]
pub fn save_recipe_cost_snapshot(
    db: State<DbState>,
    recipe_id: i64,
    pricing_strategy_summary: String,
    total_cost_micros: i64,
    cost_per_yield_unit_micros: i64,
    breakdown_json: String,
) -> AppResult<i64> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::save_cost_snapshot(
        &conn,
        recipe_id,
        &pricing_strategy_summary,
        total_cost_micros,
        cost_per_yield_unit_micros,
        &breakdown_json,
    )
}

#[tauri::command]
pub fn list_recipe_cost_snapshots(
    db: State<DbState>,
    recipe_id: i64,
) -> AppResult<Vec<CostSnapshotSummary>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    recipes::list_cost_snapshots(&conn, recipe_id)
}
